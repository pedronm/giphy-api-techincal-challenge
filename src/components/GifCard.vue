<template>
  <div class="gif-card">
    <img :alt="data.title" :src="data.images.preview_gif.url" />
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from "vue-property-decorator";

@Component
export default class GifGrid extends Vue {
  @Prop() private data!: any;
}
</script>

<style scoped lang="scss">
@import "@/styles/components/gifcard.scss";
</style>
