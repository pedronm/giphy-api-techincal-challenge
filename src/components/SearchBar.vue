<template>
    <div class="search-bar">
        
        <form class="search-bar__form" v-on:submit="triggerSearch">
            <div class="block">
                <input id="search-bar" 
                    :value="query"
                    list="suggestions"
                    type="text"
                    autocomplete="off" 
                    @input="updateQuery"/>
                <button type="submit">
                    <img src="https://giphy.com/static/img/search-icon.svg"/>
                </button>
            </div>
            <datalist v-show="suggestedQueries && suggestedQueries.length" id="suggestions">
                <option v-for="(suggestion, index) of suggestedQueries"
                    v-bind:key="index"
                    :value="suggestion.name"/>
            </datalist>
        </form>
        
    </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from "vue-property-decorator";
import store from "../store";

@Component
export default class SearchBar extends Vue {

    @Prop() private msg!: string;
    
    get query() {
        return store.getters.getQuery()
    }

    get suggestedQueries() {
        return store.getters.getSuggested()
    }

    private updateQuery = (e) => {
        store.commit('setQuery', e.target.value)
        store.dispatch('fetchSuggested');
    }

    private triggerSearch() {
        store.dispatch('fetchSearch');
    }
}
</script>

<style scoped lang="scss">

.search-bar {
    height: 2em;

    #back {
        background: white;
        height: 2.3em;
        position: relative;
        width: 67%;
        top: 2.5em;
        display: inline-block;
    }

    .block {
        height: 100%;
        display: block;
        width: 100%;
    }

    &__form {
        height: 100%;
        position: relative;

        input {
            background: #ffffff;
            height: 100%;
            width: 67%;
            border: none;
            text-indent: 1em;
            font-size: 1em;
            text-transform: lowercase;
        }

        button {
            background-color: #2e8ca5;
            display: inline-block;
            height: 106%;
            width: 3em;
            position: absolute;
            border: none;
            padding: 2px;
        }
    }
}
</style>
