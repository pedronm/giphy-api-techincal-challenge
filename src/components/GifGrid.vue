<template>
  <div class="gif-grid">
    <header>
      <h1>{{ title }}</h1>
      <search-bar v-if="contentType === 'searchResults'"
        />
      <p v-else>{{ msg }}</p>
    </header>
    <div class="image-grid" 
        :id="`grid-${contentType}`"
        v-on:scroll="scrollHandler">
        <gif-card v-for="(img, index) of images" v-bind:key="index" :data="img" />
        <div class="end-marker" 
            @click="nextBatch">
            <div v-show="loading || busy" class="loading">Loading...</div>
            <div v-show="!loading && !busy"
                class="more"
                id="more">
                <span>{{ images.length }}</span>Results
            </div>
        </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from "vue-property-decorator";

import GifCard from "@/components/GifCard.vue";
import SearchBar from "@/components/SearchBar.vue";
import store from "../store";

@Component({
  components: {
    GifCard,
    SearchBar
  }
})
export default class GifGrid extends Vue {
  private busy: boolean;

  constructor() {
    super();
    this.busy = false;
  }

  @Prop() private title!: string;
  @Prop() private msg!: string;
  @Prop() private contentType!: string;

  get images(): any[] {
    let images: any[] = [];
    switch (this.contentType) {
      case "mostPopular":
        images = store.state.mostPopular;
        /*
        // uncomment to enable auto refetch of cleaned data

        if (!images.length) {
          store.dispatch('fetchMostPopular');
        }
        */
        break;
      case "searchResults":
        images = store.state.searchResults;
        break;
      case "random":
        images = store.state.random;
        break;
    }
    return images;
  }

  get loading(): boolean {
    return store.getters.checkLoading(this.contentType);
  }


  private nextBatch() {
    this.busy = true;
    setTimeout(() => {
      this.busy = false;
    }, 1000);
    // as we consume the current next batch of results
    // then we should start fetching a subsequent batch now
    store.dispatch("fetchNextRandomBatch");
  }

  private scrollHandler = (e) => {
      const container: HTMLElement = e.target;
      if (container.scrollTop + container.clientHeight >= container.scrollHeight) {
          this.nextBatch();
      }
  }
}
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
h3 {
  margin: 40px 0 0;
}
ul {
  list-style-type: none;
  padding: 0;
}
li {
  display: inline-block;
  margin: 0 10px;
}
a {
  color: #42b983;
}

header {
  height: 20%;
  padding-top: 0.3em;
}

.gif-card {
    background-color: #ece7cb;
    box-shadow: 5px 5px 2px rgba(0, 0, 0, 0.44);
    border-radius: 10px;
    padding: 10px 10px 10px 10px;
    display: flex;
    width: auto;
}

.image-grid {
    height: 70%;
    overflow-y: scroll;
    display: grid;
    grid-template-columns: repeat(4, 1fr);
    background-color: #312f2f;
    padding: 10px;
    width: 95%;
    margin: auto;
    grid-gap: 15px;
    box-shadow: -5px -5px #151515f2;
    grid-auto-rows: min-content;
}

.gif-grid {
    height: 100%;
    background-color: #1f1a1a;
    color: #d8d3b8;
}

#more span {
    margin-right: 0.4em;
}

.end-marker {
    color: #444340;
    box-shadow: 5px 5px 2px rgba(0, 0, 0, 0.12);
    border-radius: 10px;
    padding: 10px 10px 10px 10px;
    display: flex;
    width: auto;
    background-color: #73737314;

    * {
        margin: auto;
    }
}
</style>
