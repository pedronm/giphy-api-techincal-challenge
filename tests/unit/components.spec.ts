import { shallowMount } from "@vue/test-utils";
import GifCard from "@/components/GifCard.vue";
import GifGrid from "@/components/GifGrid.vue";
import SearchBar from "@/components/SearchBar.vue";
import * as testData from "./test-data.js";

describe("GifCard.vue", () => {
  it("renders props.data when passed", () => {
    const data = testData.default.data[0];
    const wrapper = shallowMount(GifCard, {
      propsData: { data }
    });
    wrapper.html()
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("GifGrid.vue", () => {
  it("renders when passed props for: 'Most Popular'", () => {
    const title = "Most Popular";
    const msg = "Here are some of the most popular GIFs available on GIPHY right now!";
    const contentType = "mostPopular";

    const wrapper = shallowMount(GifGrid, {
      propsData: {
        title,
        msg,
        contentType,
      }
    });
    wrapper.html()
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders when passed props for: 'Search'", () => {
    const title="Search";
    const msg="Search for GIFs on GIPHY!";
    const contentType="searchResults";

    const wrapper = shallowMount(GifGrid, {
      propsData: {
        title,
        msg,
        contentType,
      }
    });
    wrapper.html()
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders when passed props for: 'Random'", () => {
    const title="Random";
    const msg="Here's a random collection of GIFs from GIPHY!";
    const contentType="random";

    const wrapper = shallowMount(GifGrid, {
      propsData: {
        title,
        msg,
        contentType,
      }
    });
    wrapper.html()
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("SearchBar.vue", () => {
  it("renders", () => {
    const wrapper = shallowMount(SearchBar, {});
    wrapper.html()
    expect(wrapper.html()).toMatchSnapshot();
  });
});
