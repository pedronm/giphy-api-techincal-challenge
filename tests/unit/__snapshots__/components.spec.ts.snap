// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifCard.vue renders props.data when passed 1`] = `<div class="gif-card"><img alt="ryan gosling GIF" src="https://media1.giphy.com/media/qg5pk8s2h5kJy/giphy-preview.gif?cid=9e3f09c8914b506df656bf82215c379b7ebde7ad89ac1987&amp;rid=giphy-preview.gif"></div>`;

exports[`GifGrid.vue renders when passed props for: 'Most Popular' 1`] = `
<div class="gif-grid">
  <header>
    <h1>Most Popular</h1>
    <p>Here are some of the most popular GIFs available on GIPHY right now!</p>
  </header>
  <div id="grid-mostPopular" class="image-grid">
    <div class="end-marker">
      <div class="loading">Loading...</div>
      <div id="more" class="more" style="display: none;"><span>0</span>Results
      </div>
    </div>
  </div>
</div>
`;

exports[`GifGrid.vue renders when passed props for: 'Random' 1`] = `
<div class="gif-grid">
  <header>
    <h1>Random</h1>
    <p>Here's a random collection of GIFs from GIPHY!</p>
  </header>
  <div id="grid-random" class="image-grid">
    <div class="end-marker">
      <div class="loading">Loading...</div>
      <div id="more" class="more" style="display: none;"><span>0</span>Results
      </div>
    </div>
  </div>
</div>
`;

exports[`GifGrid.vue renders when passed props for: 'Search' 1`] = `
<div class="gif-grid">
  <header>
    <h1>Search</h1>
    <search-bar-stub></search-bar-stub>
  </header>
  <div id="grid-searchResults" class="image-grid">
    <div class="end-marker">
      <div class="loading" style="display: none;">Loading...</div>
      <div id="more" class="more"><span>0</span>Results
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchBar.vue renders 1`] = `
<div class="search-bar">
  <form class="search-bar__form">
    <div class="block"><input id="search-bar" list="suggestions" type="text" autocomplete="off"> <button type="submit"><img src="https://giphy.com/static/img/search-icon.svg"></button></div> <datalist id="suggestions" style="display: none;"></datalist>
  </form>
</div>
`;
